### PR TITLE
prometheus-dmarc-exporter: don't use poetry2nix

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
~https://github.com/NixOS/nixpkgs/pull/174629#issuecomment-1138426280 and it currently downloads some wheels from PyPI~

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adisbladis** [comment]
The evaluation error in https://github.com/NixOS/nixpkgs/pull/174629#issuecomment-1138426280 was caused by a legitimate incompatibility where upstream had marked the package as incompatible with 3.10+.

Additionally there are no wheels downloaded from PyPi for this package.

**Comment by dotlambda** [comment]
> Additionally there are no wheels downloaded from PyPi for this package.

So how are https://pypi.org/project/dataclasses-serialization/ and https://pypi.org/project/more-properties/ installed?

**Comment by dotlambda** [comment]
> > Additionally there are no wheels downloaded from PyPi for this package.
> 
> So how are https://pypi.org/project/dataclasses-serialization/ and https://pypi.org/project/more-properties/ installed?

Oh I see they are overriden manually.

**Comment by dotlambda** [comment]
I think it should at least be up to the maintainer @Ma27 to decide whether to use poetry2nix.
However I'd argue that we should strive to avoid poetry2nix in Nixpkgs because it
* creates more load for Hydra and uses more disk space
* delays security updates
* negates the intermediate step of distro maintainers vetting software

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by adisbladis** [comment]
> creates more load for Hydra and uses more disk space

Marginally so, but sure.
I do wonder how much extra Hydra load and disk space the regular python nixpkgs development workflow adds though. I'd wager a whole lot more.

> delays security updates

How does this delay security updates exactly?

> negates the intermediate step of distro maintainers vetting software

I would argue that this is already not happening and therefore a moot argument.


But I think the biggest argument in favour of keeping such approaches around is developer time and correctness.
With these types of overrides https://github.com/NixOS/nixpkgs/pull/174804/files#diff-33ee5643df27fb56c964dcde0e3b5a50ce360c14eaae09efa1c9f62e9247a7fcR24-R28 you're sacrificing correctness in a very hacky and questionable manner.

**Comment by dotlambda** [comment]
> I do wonder how much extra Hydra load and disk space the regular python nixpkgs development workflow adds though. I'd wager a whole lot more.

A lot. But we get a properly tested package set.

> How does this delay security updates exactly?

Say a security update of dataclasses-serialization is released. The maintainer of dmarc-metrics-exporter takes a few days to update the version of dataclasses-serialization in poetry.lock (if they even hear about the security update). The Nixpkgs maintainer takes a few days to update dmarc-metrics-exporter.
It gets even worse with more indirection and packages needlessly pinning their dependencies' versions.

I think it's much better to ignore pins and properly test instead so that updates reach all reverse dependencies as soon as the Nixpkgs maintainer or @r-ryantm notices.



> I would argue that this is already not happening and therefore a moot argument.

I agree.


> But I think the biggest argument in favour of keeping such approaches around is developer time and correctness. With these types of overrides https://github.com/NixOS/nixpkgs/pull/174804/files#diff-33ee5643df27fb56c964dcde0e3b5a50ce360c14eaae09efa1c9f62e9247a7fcR24-R28 you're sacrificing correctness in a very hacky and questionable manner.

That particular override is "endorsed" by upstream: https://github.com/jgosmann/bite-parser/blob/main/pyproject.toml#L12

But in general I think that for the reasons above ignoring pins and running as many tests as possible is the better choice. Why else would you use a distro and not just `pip install` directly?

**Comment by SuperSandro2000** [comment]
> How does this delay security updates exactly?

Lets be realistic here: If upstream is not releasing as often as hypothesis is then a python lock file gets updated eventually and is released with the next update and that could be quite a while or in some cases never.

Also we do not want to land in the pitty Go and Rust are already in where you need to check every single package for security related updates which looks like https://github.com/NixOS/nixpkgs/issues/141368

> With these types of overrides [#174804 (files)](https://github.com/NixOS/nixpkgs/pull/174804/files#diff-33ee5643df27fb56c964dcde0e3b5a50ce360c14eaae09efa1c9f62e9247a7fcR24-R28) you're sacrificing correctness in a very hacky and questionable manner.

Widening version constraints is something that is constantly done and required in pythonPackages. Otherwise we couldn't deliver some software at all due to the state of the python ecosystem. Generally all core python maintainers treat version constraints a bit lightly. If there are tests for the package more often than not a version constraint is removed.

Also that particular version constraint should be a noop and could probably be dropped.

> Why else would you use a distro and not just pip install directly?

Because pip is not very correct at all. It happily changes install order of packages which also can effect version constraints and up until recently pip sometimes choose the wrong wheels on alpine. It also can't track which packages you installed and which ones are dependencies, ignores version constraints on future operations and if you supply it with an impossible version constraint scenario you can end up downloading and trying to build every version of every dependency in your entire dependency chain.

**Comment by dotlambda** [comment]
> > Why else would you use a distro and not just pip install directly?
> 
> Because pip is not very correct at all.

I should have said `poetry install`, not `pip install`.

**Comment by Ma27** [comment]
Back when I packaged it, I considered packaging each dependency to be too much of a pain, so I went down the poetry2nix route (which seemed like an OK choice given that it was already used in nixpkgs).

Given that the Python maintainers seem to prefer to use python packages and all dependencies are packaged now, it seems OK to me.

Also, LGTM implementation-wise.

**Comment by dotlambda** [comment]
> The evaluation error in [#174629 (comment)](https://github.com/NixOS/nixpkgs/pull/174629#issuecomment-1138426280) was caused by a legitimate incompatibility where upstream had marked the package as incompatible with 3.10+.

It's not uncommon for poetry2nix to break eval: https://github.com/NixOS/nixpkgs/pull/139480#issuecomment-1108053393
I think that's a design flaw.

File: pkgs/development/python-modules/dmarc-metrics-exporter/default.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+, typing-extensions
+, uvicorn
+, xsdata
+, aiohttp
+, pytest-asyncio
+, pytestCheckHook
+, requests
+}:
+
+buildPythonPackage rec {
```
Comment (mweinelt) [inline pkgs/development/python-modules/dmarc-metrics-exporter/default.nix:18]: Wouldn't `buildPythonApplication` and tracking it outside of pythonPackages be more suitable? It's an application after all.

File: pkgs/servers/monitoring/prometheus/dmarc-exporter/default.nix [hunk 2]
```diff
@@ -1,62 +1,14 @@
...
-    ];
-  }) python;
-  env = python.withPackages (p: [ p.dmarc-metrics-exporter ]);
+  env = python3.withPackages (p: [ p.dmarc-metrics-exporter ]);
 in
 
-(pkgs.writeShellScriptBin "prometheus-dmarc-exporter" ''
+(writeShellScriptBin "prometheus-dmarc-exporter" ''
   export PYTHONPATH="${env}/lib/${env.libPrefix}/site-packages''${PYTHONPATH:+:}''${PYTHONPATH}"
   exec ${env}/bin/python3 -m dmarc_metrics_exporter "$@"
```
Comment (mweinelt) [inline pkgs/servers/monitoring/prometheus/dmarc-exporter/default.nix:11]: Wouldn't that be covered by what `buildPythonApplication` offers?

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -2454,6 +2458,8 @@ in {
 
   dlx = callPackage ../development/python-modules/dlx { };
 
+  dmarc-metrics-exporter = callPackage ../development/python-modules/dmarc-metrics-exporter { };
```
Comment (mweinelt) [inline pkgs/top-level/python-packages.nix:2461]:
```suggestion
  dmarc-metrics-exporter = callPackage ../servers/monitoring/prometheus/dmarc-metrics-exporter { };
```

**Review by mweinelt — APPROVED** [review]

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 174804` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>dmarc-metrics-exporter</li>
    <li>python310Packages.bite-parser</li>
    <li>python310Packages.dataclasses-serialization</li>
    <li>python310Packages.docformatter</li>
    <li>python310Packages.more-properties</li>
    <li>python310Packages.xsdata</li>
    <li>python39Packages.bite-parser</li>
    <li>python39Packages.dataclasses-serialization</li>
    <li>python39Packages.docformatter</li>
    <li>python39Packages.more-properties</li>
    <li>python39Packages.xsdata</li>
  </ul>
</details>

**Pull request merged**